cloudpickle/cloudpickle_fast.py

```diff
@@ -394,95 +456,26 @@ def _class_setstate(obj, state):
-    dispatch[weakref.WeakSet] = _weakset_reduce
-    dispatch[typing.TypeVar] = _typevar_reduce
-
-    def __init__(self, file, protocol=None, buffer_callback=None):
-        if protocol is None:
-            protocol = DEFAULT_PROTOCOL
-        Pickler.__init__(self, file, protocol=protocol, buffer_callback=buffer_callback)
-        # map functions __globals__ attribute ids, to ensure that functions
-        # sharing the same global namespace at pickling time also share their
-        # global namespace at unpickling time.
-        self.globals_ref = {}
-
-        # Take into account potential custom reducers registered by external
-        # modules
-        self.dispatch_table = copyreg.dispatch_table.copy()
-        self.dispatch_table.update(self.dispatch)
-        self.proto = int(protocol)
-
-    def reducer_override(self, obj):
-        """Type-agnostic reducing callback for function and classes.
-
-        For performance reasons, subclasses of the C _pickle.Pickler class
-        cannot register custom reducers for functions and classes in the
-        dispatch_table. Reducer for such types must instead implemented in the
-        special reducer_override method.
-
-        Note that method will be called for any object except a few
-        builtin-types (int, lists, dicts etc.), which differs from reducers in
-        the Pickler's dispatch_table, each of them being invoked for objects of
-        a specific type only.
-
-        This property comes in handy for classes: although most classes are
-        instances of the ``type`` metaclass, some of them can be instances of
-        other custom metaclasses (such as enum.EnumMeta for example). In
-        particular, the metaclass will likely not be known in advance, and thus
-        cannot be special-cased using an entry in the dispatch_table.
-        reducer_override, among other things, allows us to register a reducer
-        that will be called for any class, independently of its type.
-
-
-        Notes:
-
-        * reducer_override has the priority over dispatch_table-registered
-          reducers.
-        * reducer_override can be used to fix other limitations of cloudpickle
-          for other types that suffered from type-specific reducers, such as
-          Exceptions. See https://github.com/cloudpipe/cloudpickle/issues/248
-        """
-        t = type(obj)
-        try:
-            is_anyclass = issubclass(t, type)
-        except TypeError:  # t is not a class (old Boost; see SF #502085)
-            is_anyclass = False
-
-        if is_anyclass:
-            return _class_reduce(obj)
-        elif isinstance(obj, types.FunctionType):
-            return self._function_reduce(obj)
-        else:
-            # fallback to save_global, including the Pickler's distpatch_table
-            return NotImplemented
+    # Take into account potential custom reducers registered by external
+    # modules
+    dispatch_table = copyreg.dispatch_table.copy()
```

Comment:
What happens if `copyreg` has a new override registered with it? Will this get updated?

Reply:
One thought instead of copying might be to use [`ChainMap`]( https://docs.python.org/3.8/library/collections.html#collections.ChainMap ) to combine an internal dispatch table with `copyreg`'s. That way we still pick up details from `copyreg`, but can prioritize our own mapping.

Reply:
I agree the previous code did not have this issue.